internal/enricher/geolocation.go

```diff
@@ -0,0 +1,264 @@
+package enricher
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/filemanager"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/stats"
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	"github.com/rudderlabs/rudder-server/services/geolocation"
+	"github.com/rudderlabs/rudder-server/utils/types"
+)
+
+const (
+	ERR_INVALID_IP    = "invalid_ip"
+	ERR_EMPTY_IP      = "empty_ip"
+	ERR_LOCATE_FAILED = "locate_failed"
+)
+
+type Geolocation struct {
+	IP       string `json:"ip"`
+	City     string `json:"city"`
+	Country  string `json:"country"`
+	Region   string `json:"region"`
+	Postal   string `json:"postal"`
+	Location string `json:"location"`
+	Timezone string `json:"timezone"`
+}
+
+type geoEnricher struct {
+	fetcher geolocation.GeoFetcher
+	logger  logger.Logger
+	stats   stats.Stats
+}
+
+func NewGeoEnricher(conf *config.Config, log logger.Logger, statClient stats.Stats) (PipelineEnricher, error) {
+	log.Infof("Setting up new event geo enricher")
+
+	dbPath, err := downloadMaxmindDB(context.Background(), conf, log)
+	if err != nil {
+		return nil, fmt.Errorf("downloading instance of maxmind db: %w", err)
+	}
+
+	fetcher, err := geolocation.NewMaxmindDBReader(dbPath)
+	if err != nil {
+		return nil, fmt.Errorf("creating new instance of maxmind's geolocation db reader: %w", err)
+	}
+
+	return &geoEnricher{
+		fetcher: fetcher,
+		stats:   statClient,
+		logger:  log.Child("geolocation"),
+	}, nil
+}
+
+// Enrich function runs on a request of GatewayBatchRequest which contains
+// multiple singular events from a source. The enrich function augments the
+// geolocation information per event based on IP address.
+func (e *geoEnricher) Enrich(source *backendconfig.SourceT, request *types.GatewayBatchRequest) error {
+	if !source.GeoEnrichment.Enabled {
+		return nil
+	}
+
+	e.logger.Debugw("received a call to enrich gateway events for source", "sourceID", source.ID)
+	defer e.stats.NewTaggedStat(
+		"proc_geo_enricher_request_latency",
+		stats.TimerType,
+		stats.Tags{
+			"sourceId":    source.ID,
+			"sourceType":  source.SourceDefinition.Type,
+			"workspaceId": source.WorkspaceID,
+		},
+	).RecordDuration()()
+
+	var enrichErrs []error
+	for _, event := range request.Batch {
+		// if the context section is missing on the event
+		// set it with default as map[string]interface{}
+		if _, ok := event["context"]; !ok {
+			event["context"] = map[string]interface{}{}
+		}
+
+		// if the context is other than map[string]interface{}, continue
+		context, ok := event["context"].(map[string]interface{})
+		if !ok {
+			continue
+		}
```

Comment:
I envisioned this as a missing pre condition to enrichment. The error would be when all the preconditions like `context` and `geo` within the context are present and then the enrichment fails due to some reason.

Wdyt ?

Reply:
Another point, I am thinking is we can't control the events that are being sent by the customer. If the event treats the context as some other type than map[string]interface{}, treating it as error might generate unnecessary noise in the system.

We can always mention to start with that these are pre-conditions for the event to be enriched with geo information. 

If we start treating it as an error and logging it, I am worried that we might generate logs and noise in system.

Reply:
Context is supposed to be a common field of type object, different types shouldn't be common at all
https://www.rudderstack.com/docs/event-spec/standard-events/common-fields/

Reply:
I agree but we have no checks in place to disallow for events not conforming to a specified schema, so thought of making it as a pre-condition rather than an error. It's a good point to think about and maybe in future we get more clarity to treat it as an error or otherwise.